### PR TITLE
Dev/bind algorithms candidate

### PR DESCRIPTION
**Pull request by tao558**
This is the branch dev/bind-algorithms rebased on the latest master. I'm leaving a copy of the original on my fork, as the rebase onto master was fairly involved.

### CODE REVIEW

**Comment by tao558** [comment]
Jenkins test this please

**Comment by tao558** [comment]
Jenkins test this please

**Comment by tao558** [comment]
It looks like these are failing due to the Windows path-length issue. The issue is lengthy and doesn't have a straightforward solution, so I'll explain it here.

In order to prevent collisions between python modules, we are using longer module names (usually stored in safe_modpath in the CMake code). These long path names are essentially repeated by MSBuild when generating the .lastbuildstate files. When the configuration files are generated, the string "configure-" is added to the path, only making it longer, which is actually causing the issue in this case. We'd need some way to shorten the safe_modpath in order to get this passing, while still guaranteeing no collisions.

The call to generate the configure files is here: https://github.com/Kitware/kwiver/blob/0a35ab88d5ea6932961e713548c3f9b8780fdadb/CMake/utils/kwiver-utils-python.cmake#L167

The actual generation of configuration files occurs in this function: https://github.com/Kitware/kwiver/blob/0a35ab88d5ea6932961e713548c3f9b8780fdadb/CMake/utils/kwiver-utils-configuration.cmake#L38

What are your thoughts? @as6520

**Comment by Erotemic** [comment]
Uggg, windows path length issues. That is one of the top reasons I switched to Linux. 

In the past I've done really hacky things to get around it like hashing part of the string to shorten it.
I've done things where I find minimum unique prefixes and other crazy things like that.  I actually do the first thing in [ubelt](https://github.com/Erotemic/ubelt/blob/master/ubelt/util_cache.py#L225) also for the win32 case: 

```
    def _condense_cfgstr(self, cfgstr=None):
        cfgstr = self._rectify_cfgstr(cfgstr)
        # The 49 char maxlen is just long enough for an 8 char name, an 1 char
        # underscore, and a 40 char sha1 hash.
        max_len = 49
        if len(cfgstr) > max_len:
            from ubelt import util_hash
            condensed = util_hash.hash_data(cfgstr, hasher=self.hasher,
                                            base='hex')
            condensed = condensed[0:max_len]
        else:
            condensed = cfgstr
        return condensed
```

Due to my suborn insistence of having support for win32 in all ubelt functions, I have a good amount of hacky win32 solutions: e.g. for [almost-symlinks](https://github.com/Erotemic/ubelt/blob/master/ubelt/_win32_links.py) on windows. 

I don't know if the hashing solution is appropriate here. Obviously the python module names will need to be preserved, but maybe we can hash any prefixes?

**Comment by tao558** [comment]
> Uggg, windows path length issues. That is one of the top reasons I switched to Linux.
> 
> In the past I've done really hacky things to get around it like hashing part of the string to shorten it.
> I've done things where I find minimum unique prefixes and other crazy things like that. I actually do the first thing in [ubelt](https://github.com/Erotemic/ubelt/blob/master/ubelt/util_cache.py#L225) also for the win32 case:
> 
> ```
>     def _condense_cfgstr(self, cfgstr=None):
>         cfgstr = self._rectify_cfgstr(cfgstr)
>         # The 49 char maxlen is just long enough for an 8 char name, an 1 char
>         # underscore, and a 40 char sha1 hash.
>         max_len = 49
>         if len(cfgstr) > max_len:
>             from ubelt import util_hash
>             condensed = util_hash.hash_data(cfgstr, hasher=self.hasher,
>                                             base='hex')
>             condensed = condensed[0:max_len]
>         else:
>             condensed = cfgstr
>         return condensed
> ```
> 
> Due to my suborn insistence of having support for win32 in all ubelt functions, I have a good amount of hacky win32 solutions: e.g. for [almost-symlinks](https://github.com/Erotemic/ubelt/blob/master/ubelt/_win32_links.py) on windows.
> 
> I don't know if the hashing solution is appropriate here. Obviously the python module names will need to be preserved, but maybe we can hash any prefixes?

I may have fixed the issue by renaming algorithms to algos. This may seem like a small change, but we were barely over in the first place, and that part of the path is repeated. We'll have to see how the windows build looks

**Comment by tao558** [comment]
Forgot to change it in the tests directory

**Comment by tao558** [comment]
Jenkins test this please

**Comment by tao558** [comment]
Jenkins test this please

**Comment by tao558** [comment]
Last fix resolved all path-length issues except 2. This should fix those last 2

**Comment by dstoup** [comment]
Looks good to me. Just make sure to address the ghostflow-check errors. Thanks!

**Comment by tao558** [comment]
Jenkins test this please

**Comment by tao558** [comment]
Jenkins test this please

File: CMake/utils/kwiver-utils-python.cmake
Comment (as6520) [inline CMake/utils/kwiver-utils-python.cmake:152]: Where is this variable used?

File: python/kwiver/vital/algo/algorithm.h
Comment (as6520) [inline python/kwiver/vital/algo/algorithm.h:58]: Should this modified function signature be used in `get_nested_algo_configuration`?

File: python/kwiver/vital/algo/algorithm_factory.cxx
Comment (as6520) [inline python/kwiver/vital/algo/algorithm_factory.cxx:55]: Where is this function being used?

File: CMake/utils/kwiver-utils-python.cmake
Comment (tao558) [reply CMake/utils/kwiver-utils-python.cmake:152]: I don't believe it is. Can you confirm, @johnwparent?

File: python/kwiver/vital/algo/algorithm_factory.cxx
Comment (tao558) [reply python/kwiver/vital/algo/algorithm_factory.cxx:55]: In the bindings below: https://github.com/Kitware/kwiver/blob/57cec13138c07c9b2de04edb6bb8f72975e054e5/python/kwiver/vital/algo/algorithm_factory.cxx#L96

And in some tests: https://github.com/Kitware/kwiver/blob/57cec13138c07c9b2de04edb6bb8f72975e054e5/python/kwiver/vital/tests/test_algorithm_factory.py#L84

https://github.com/Kitware/kwiver/blob/57cec13138c07c9b2de04edb6bb8f72975e054e5/python/kwiver/vital/tests/test_algorithms_common.py#L146

File: CMake/utils/kwiver-utils-python.cmake
Comment (johnwparent) [reply CMake/utils/kwiver-utils-python.cmake:152]: This variable is used to pass the final location of the tests module to nose. Used in kwiver-setup-tests-python.cmake line 37.

File: python/kwiver/vital/algo/algorithm_factory.cxx
Comment (as6520) [reply python/kwiver/vital/algo/algorithm_factory.cxx:55]: Wouldn't the namespace be kwiver::vital::python then? I was confused by the use kwiver::vital:: namespace.

File: CMake/utils/kwiver-utils-python.cmake
Comment (johnwparent) [reply CMake/utils/kwiver-utils-python.cmake:152]: Yes, it absolutely should.

File: python/kwiver/vital/algo/algorithm_factory.cxx
Comment (tao558) [reply python/kwiver/vital/algo/algorithm_factory.cxx]: Oh no you are correct. This is actually not used, as I thought I would need to write an entire new create_object function, when instead the one already defined in kwiver::vital can be used. Pushed a fix in abd3f1fdda1530772d6e50db05e4e2b594c950c1

**Comment by tao558** [comment]
Jenkins test this please

**Comment by tao558** [comment]
Jenkins test this please

File: CMake/utils/kwiver-utils-python.cmake
Comment (tao558) [reply CMake/utils/kwiver-utils-python.cmake]: This commit was removed in favor of a branch with all of the nose integration: https://github.com/Kitware/kwiver/tree/dev/add_nose. All that should be necessary after that branch is merged into master is a rebase.

**Comment by dstoup** [comment]
I can't speak to the validity of the python bindings but I don't see anything that stands out as bad. Would be great to get a review from someone who knows the binding process better, but it looks good to me.

File: python/kwiver/vital/algo/estimate_fundamental_matrix.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/estimate_fundamental_matrix.cxx:46]: Please add class-level and method-level docstrings, ideally with doctest example blocks.

File: python/kwiver/vital/algo/estimate_fundamental_matrix.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/estimate_fundamental_matrix.cxx:66]: I often find that overloaded C bindings can be a pain when trying to get the right signature. Is there a reason to expose both variants of the estimate method?

My `C` syntax understanding is less than stellar. I'm slightly confused by the nested parens without any commas between what appear to be items. Can you break down what the syntax here is saying? 

```C++
( kwiver::vital::fundamental_matrix_sptr
           (kwiver::vital::algo::estimate_fundamental_matrix::* )
           ( const kwiver::vital::feature_set_sptr,
             const kwiver::vital::feature_set_sptr,
             const kwiver::vital::match_set_sptr,
             std::vector< bool >&,
             double ) const 
```

File: python/kwiver/vital/algo/estimate_homography.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/estimate_homography.cxx:46]: A Python API needs docstrings to maximize usability. A developer should always be able to ask what a class or method does in the code itself.

File: python/kwiver/vital/algo/estimate_pnp.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/estimate_pnp.cxx:45]: Docstrings

File: python/kwiver/vital/algo/estimate_similarity_transform.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/estimate_similarity_transform.cxx:48]: docstrings

File: python/kwiver/vital/algo/extract_descriptors.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/extract_descriptors.cxx:45]: docstrings

File: python/kwiver/vital/algo/feature_descriptor_io.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/feature_descriptor_io.cxx:47]: docstrings

File: python/kwiver/vital/algo/filter_features.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/filter_features.cxx:46]: docstrings

File: python/kwiver/vital/algo/filter_tracks.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/filter_tracks.cxx:46]: docstrings

File: python/kwiver/vital/algo/image_filter.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/image_filter.cxx:45]: docstrings

File: python/kwiver/vital/algo/image_io.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/image_io.cxx:51]: docstrings

File: python/kwiver/vital/algo/initialize_cameras_landmarks.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/initialize_cameras_landmarks.cxx:48]: docstrings

File: python/kwiver/vital/algo/integrate_depth_maps.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/integrate_depth_maps.cxx:46]: docstrings

File: python/kwiver/vital/algo/interpolate_track.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/interpolate_track.cxx:46]: docstrings

File: python/kwiver/vital/algo/keyframe_selection.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/keyframe_selection.cxx:46]: docstrings

File: python/kwiver/vital/algo/match_descriptor_sets.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/match_descriptor_sets.cxx:46]: docstrings

File: python/kwiver/vital/algo/match_features.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/match_features.cxx:47]: docstrings

File: python/kwiver/vital/algo/optimize_cameras.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/optimize_cameras.cxx:46]: docstrings

File: python/kwiver/vital/algo/read_object_track_set.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/read_object_track_set.cxx:45]: docstrings

File: python/kwiver/vital/algo/read_track_descriptor_set.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/read_track_descriptor_set.cxx:45]: docstrings

File: python/kwiver/vital/algo/refine_detections.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/refine_detections.cxx:45]: docstrings

File: python/kwiver/vital/algo/split_image.cxx
Comment (Erotemic) [inline python/kwiver/vital/algo/split_image.cxx:45]: docstrings

File: python/kwiver/vital/tests/alg/simple_match_features.py
Comment (Erotemic) [inline python/kwiver/vital/tests/alg/simple_match_features.py:37]: All this algo register stuff is unwieldy. I have a proposal for a better alternative scheme here: #1123

**Review by Erotemic — CHANGES_REQUESTED** [review]

**Comment by Erotemic** [comment]
In my review, I basically want docstrings in every Python binding. After a certain point, I stopped commenting in each module, but each and every python binding should have a docstring specifying: A one sentence high-level overview, optionally one or two paragraphs with more detail, preferably a Args and Returns block detailing the arguments and return types, and ideally a doctest demonstrating example usage (and providing test coverage via xdoctest). 

I also have a comment wrt to vital registration #1123 and copyright headers #1014

File: CMake/kwiver-setup-tests-python.cmake
Comment (johnwparent) [reply CMake/kwiver-setup-tests-python.cmake]: Switching everything over to pytest is one of my priorities after #1076 is merged

**Comment by mleotta** [comment]
I agree that docstrings are important, but I also expect that adding them is a big undertaking, especially since the documentation may not already be there on the C++ side, or if it does, it may not exist in a consistent format.

My recommendation would be to get this branch merged with without the docstrings, or with some generic placeholder docstring for now.  Then we need to come up with a standard way of providing documentation strings that will show up both in the C++ doxygen documentation and in the Python docstring.  Once we have mechanism for doing that we need to go back add the documentation in that format and pull on the original authors of these functions to help write that content.  It's very important to do this, but I'm in favor of a staged approach.  Land this branch first and then address documentation in a follow-up.

File: python/kwiver/vital/algo/estimate_canonical_transform.cxx
Comment (johnwparent) [reply python/kwiver/vital/algo/estimate_canonical_transform.cxx:50]: An AutoPybind11 feature to automatically extract c++ comments and feed them into the pybind doc strings is currently under development. In the meantime, pybind provides a tool, [mkdoc](https://github.com/pybind/pybind11_mkdoc) that could be used to accomplish this.

**Comment by Erotemic** [comment]
@mleotta I agree with that plan, but I will note that I'm afraid that putting it off means it will never get done. 

At the very least, I do want the docstring placeholder in every binding, so it is clear where the documentation should go.

**Comment by mleotta** [comment]
I agree with that plan.  We need to do what we can on this branch to make it easier to get the documentation done later and maybe some incentive or reminder to do so.  We could have a generic placeholder docstring that says "documentation missing" or something like that.  Then we can ask all KWIVER developers to go in a fill in the missing documentation.  We can search for this placeholder string to easily identify which cases have not been addressed.

**Review by dstoup — APPROVED** [review]
I'm ok with this PR. There is the discussion of Docstrings to resolve but I'm ok with waiting until the feature to generate them is added to auto-pybind11.

I will let others with a more passionate feelings make the final call.

**Pull request merged**
